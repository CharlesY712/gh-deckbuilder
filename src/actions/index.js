export const addCard = (card) => ({
  type: 'ADD_CARD',
  card
});

export const addCards = (cards) => ({
  type: 'ADD_CARDS',
  cards
});

export const removeCards = (cards) => ({
  type: 'REMOVE_CARDS',
  cards
});

export const addSelectedCard = (selectedCard) => ({
  type: 'ADD_SELECTED_CARD',
  selectedCard
});

export const addSelectedCards = (selectedCards) => ({
  type: 'ADD_SELECTED_CARDS',
  selectedCards
});

export const removeSelectedCard = (selectedCard) => ({
  type: 'REMOVE_SELECTED_CARD',
  selectedCard
});

export const removeSelectedCards = (selectedCards) => ({
  type: 'REMOVE_SELECTED_CARDS',
  selectedCards
});

export const addAvailableCard = (availableCard) => ({
  type: 'ADD_AVAILABLE_CARD',
  availableCard
});

export const addAvailableCards = (availableCards) => ({
  type: 'ADD_AVAILABLE_CARDS',
  availableCards
});

export const removeAvailableCard = (availableCard) => ({
  type: 'REMOVE_AVAILABLE_CARD',
  availableCard
});

