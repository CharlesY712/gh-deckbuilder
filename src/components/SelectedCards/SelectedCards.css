/* Most of the styles for this component reside in AvailableCards.css */