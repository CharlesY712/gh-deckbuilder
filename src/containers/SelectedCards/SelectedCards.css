#selected-component {
  width: 42%;
  margin-left: 20px;
}

#selected-heading {
  text-align: right;
  right: 8%;
}

/* Most of the styles for this component reside in AvailableCards.css */