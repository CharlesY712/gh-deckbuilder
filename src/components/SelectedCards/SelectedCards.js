import React from 'react';
import { connect } from 'react-redux';
import './SelectedCards.css';
import { Card } from '../Card/Card'
import {
  addSelectedCard,
  removeAvailableCard
} from '../../actions';

export const SelectedCards = ({ cards }) => {
  let displayCards

  if(cards) {
    displayCards = cards.map(card => {
      return <Card
        key={card.id}
        id={card.id}
        image={card.image_url}
        name={card.name} />;
    });
  }

  const dragoverHandler = (event) => {
    event.preventDefault();
    //make shit move out of the way
    event.dataTransfer.dropEffect = "move"
  }

  const dropHandler = (event) => {
    event.preventDefault();
    //don't allow to drop on original parent
    var data = event.dataTransfer.getData("text");
    console.log(data)
    //removeAvailableCard
    //addSelectedCard
    event.target.appendChild(document.getElementById(data));
  }

  return (
    <div className="cards-component" id="selected-component">
      <h2>Selected Cards/Deck Name</h2>
      <div className="cards-container"
        onDrop={event => dropHandler(event)}
        onDragOver={event => dragoverHandler(event)} >
        {displayCards}
      </div>
    </div>
  );
}

//make sure these actions are working!
export const mapDispatchToProps = dispatch => ({
  removeAvailableCard: availableCard =>
    dispatch(removeAvailableCard(availableCard)),
  addSelectedCard: selectedCard =>
    dispatch(addSelectedCard(selectedCard))
});

export const mapStateToProps = state => ({
  cards: state.cards,
  selectedCards: state.selectedCards,
  availableCards: state.availableCards
});

export default connect(mapStateToProps, mapDispatchToProps)(SelectedCards);
