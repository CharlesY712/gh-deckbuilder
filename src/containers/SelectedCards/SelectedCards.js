import React, {Component} from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
import './SelectedCards.css';
import * as helpers from '../../helpers/index';
import * as actions from '../../actions';

export class SelectedCards extends Component {

  render() {

    const { cards, selectedCards, selectedClass, currentLevel,
      addSelectedCard, removeAvailableCard } = this.props;

    const dragoverHandler = (event) => {
      event.preventDefault();
      event.dataTransfer.dropEffect = "move";
    };

    const dropHandler = (event) => {
      event.preventDefault();
      const data = JSON.parse(event.dataTransfer.getData("text"));
      const id = data.id;
      const parent = data.parent;
      const handSize = selectedCards.length;
      const handLimit = helpers.getHandSize(selectedClass);
      if (parent === "available-component" && handSize < handLimit) {
        const droppedCard = helpers.getCardById(parseInt(id, 10), cards);
        addSelectedCard(droppedCard);
        removeAvailableCard(droppedCard);
      } 
    };

    return (
      <div className="cards-component" id="selected-component">
        <h2 id="selected-heading">Selected Cards</h2>
        <div className="scroll-fade-top"></div>
        <div className="cards-container"
          onDrop={event => dropHandler(event)}
          onDragOver={event => dragoverHandler(event)} >
          {helpers.renderCards(selectedCards, currentLevel)}
        </div>
        <div className="scroll-fade-bottom"></div>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  cards: state.cards,
  selectedCards: state.selectedCards,
  availableCards: state.availableCards,
  selectedClass: state.selectedClass,
  currentLevel: state.currentLevel
});

export const mapDispatchToProps = dispatch => ({
  removeAvailableCard: availableCard =>
    dispatch(actions.removeAvailableCard(availableCard)),
  addSelectedCard: selectedCard =>
    dispatch(actions.addSelectedCard(selectedCard))
});

SelectedCards.propTypes = {
  cards: PropTypes.array,
  selectedCards: PropTypes.array,
  selectedClass: PropTypes.string,
  currentLevel: PropTypes.number,
  addSelectedCard: PropTypes.func,
  removeAvailableCard: PropTypes.func
};

export default withRouter(connect(mapStateToProps,
  mapDispatchToProps)(SelectedCards));
