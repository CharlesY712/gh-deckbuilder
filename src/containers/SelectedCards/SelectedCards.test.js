import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { SelectedCards } from './SelectedCards';
import { mapStateToProps, mapDispatchToProps } from './SelectedCards';
import * as actions from '../../actions';
import * as mocks from '../../mocks/mockCards';

describe("Selected Card component", () => {
  describe("Container Tests", () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<SelectedCards
        addSelectedCard={jest.fn()}
        availableCards={[]}
        cards={[]}
        currentLevel={1}
        removeAvailableCard={jest.fn()}
        selectedCards={[]}
        selectedClass={"Brute"}
      />)
    });

    it("should match the snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });

  });

  describe("Map State to Props", () => {
    it('correctly maps cards to props', () => {
      const mockState = {cards: mocks.mockCards, selectedCards: [], availableCards: [], selectedClass: "", currentLevel: 0};
      const mapped = mapStateToProps(mockState);
      const expectedState = mocks.mockCards;

      expect(mapped.cards).toEqual(expectedState);
    });

    it("correctly maps selectedCards to props", () => {
      const mockState = {cards: [], selectedCards: mocks.mockSelectedCards, availableCards: [], selectedClass: "", currentLevel: 0};
      const mapped = mapStateToProps(mockState);
      const expectedState = mocks.mockSelectedCards;

      expect(mapped.selectedCards).toEqual(expectedState);
    });

    it("correctly maps availableCards to props", () => {
      const mockState = {cards: [], selectedCards: mocks.mockSelectedCards, availableCards: mocks.mockAvailableCards, selectedClass: "", currentLevel: 0};
      const mapped = mapStateToProps(mockState);
      const expectedState = mocks.mockAvailableCards;

      expect(mapped.availableCards).toEqual(expectedState);
    });

    it("correctly maps selectedClass to props", () => {
      const mockState = {cards: [], selectedCards: [], availableCards: [], selectedClass: "Brute", currentLevel: 0};
      const mapped = mapStateToProps(mockState);
      const expectedState = "Brute";

      expect(mapped.selectedClass).toEqual(expectedState);
    });

    it("correctly maps currentLevel to props", () => {
      const mockState = {cards: [], selectedCards: [], availableCards: [], selectedClass: "", currentLevel: 4};
      const mapped = mapStateToProps(mockState);
      const expectedState = 4;

      expect(mapped.currentLevel).toEqual(expectedState);
    });

  });

  describe("Map Dispatch to Props", () => {
    let mockDispatch;
    let mappedDispatch;

    beforeEach(() => {
      mockDispatch = jest.fn();
      mappedDispatch = mapDispatchToProps(mockDispatch);
    });

    it("Should dispatch removeAvailableCard", () => {
      mappedDispatch.removeAvailableCard();
      const expected = actions.removeAvailableCard();
      expect(mockDispatch).toHaveBeenCalledWith(expected);
    });

    it("should dispatch addSelectedCard", () => {
      mappedDispatch.addSelectedCard();
      const expected = actions.addSelectedCard();
      expect(mockDispatch).toHaveBeenCalledWith(expected);
    });
  });
});
