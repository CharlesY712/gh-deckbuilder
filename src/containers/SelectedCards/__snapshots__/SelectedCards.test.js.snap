// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selected Card component Container Tests should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectedCards
    addSelectedCard={[Function]}
    availableCards={Array []}
    cards={Array []}
    currentLevel={1}
    removeAvailableCard={[Function]}
    selectedCards={Array []}
    selectedClass="Brute"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          id="selected-heading"
>
          Selected Cards
</h2>,
        <div
          className="scroll-fade-top"
/>,
        <div
          className="cards-container"
          onDragOver={[Function]}
          onDrop={[Function]}
>
          
</div>,
        <div
          className="scroll-fade-bottom"
/>,
      ],
      "className": "cards-component",
      "id": "selected-component",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Selected Cards",
          "id": "selected-heading",
        },
        "ref": null,
        "rendered": "Selected Cards",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "scroll-fade-top",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "cards-container",
          "onDragOver": [Function],
          "onDrop": [Function],
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "scroll-fade-bottom",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            id="selected-heading"
>
            Selected Cards
</h2>,
          <div
            className="scroll-fade-top"
/>,
          <div
            className="cards-container"
            onDragOver={[Function]}
            onDrop={[Function]}
>
            
</div>,
          <div
            className="scroll-fade-bottom"
/>,
        ],
        "className": "cards-component",
        "id": "selected-component",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Selected Cards",
            "id": "selected-heading",
          },
          "ref": null,
          "rendered": "Selected Cards",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "scroll-fade-top",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "cards-container",
            "onDragOver": [Function],
            "onDrop": [Function],
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "scroll-fade-bottom",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
